### DISCUSSION TITLE
Shall we consider making ti.static equivalent to ti.template() paramenters?

### DISCUSSION THREAD

**Discussion by k-ye** [taichi-dev/taichi#4454] [Q&A]
E.g.

```py

FLAG = True
@ti.kernel
def foo():
  if ti.static(FLAG):
    do_x()
  else:
    do_y()

foo()
FLAG = False
foo()
```

Right now this won't get re-compiled. Our general advice to workaround this is to pass `FLAG` as a `ti.template()`. However, from the users' perspective, change in the evaluated-`ti.static`-expression should trigger a re-compilation?

**Comment by strongoier** [comment]
I think this problem is not specific to `ti.static`. Any change in global variables will not trigger re-compilation, which leads to confusing behavior from time to time. We tried to solve this problem earlier by catching all global variables, however that didn't work because not all user objects are hashable (imagine that you have a numpy ndarray).
